fs: implement uv_fs_mkstemp on unix

ralt opened the pull request:
Follow-up of https://github.com/nodejs/node/issues/5332

I unfortunately don't have windows available.

The win code will have to be refactored to use `fs__mkdtemp`'s code. As far as I can see, the easiest way would be to pass a function pointer to either create a folder or a file.

I'm aware that the windows side is the most painful, code-wise. This PR is just a basis to encourage someone else to do the windows side. No bad feeling if this doesn't get done, I'd completely understand.

saghul commented:
Thanks for kickstarting this @ralt! A quick observation: we need the fd to be opened with O_CLOEXEC, so I guess we need to use `mkostemp` under the hood, at least on Linux, with a fallback on other Unix platforms (see how `uv_fs_open` does it).

As for the Windows side, we need it before we land this. For `uv_fs_mkdtemp` we borrowed inspiration from OpenBSD's libc.

ralt commented:
Hi @saghul 

Thanks for the early feedback, I'll see how to fix this.

The "function pointer" stuff I mentioned in the first post is for win. Only this bit has to change: https://github.com/libuv/libuv/blob/v1.x/src/win/fs.c#L806-L813

So passing a function pointer e.g. `int (*create)(const char* path)` would let you `break` or not in the existing `fs__mkdtemp` (after renaming it to `fs__mktemp`).

saghul commented:
> So passing a function pointer e.g. int (*create)(const char* path) would let you break or not in the existing fs__mkdtemp (after renaming it to fs__mktemp).

Ah, I see. Give it a shot and lets see how it turns out!

ralt commented:
I won't be able to even compile it, I don't have any windows available to test :-)

saghul commented:
We have a CI, don't worry.

ralt commented:
Well, I'll try blind coding then :-)

orangemocha commented:
There was some discussion about this topic here: https://github.com/libuv/libuv/issues/322

ralt commented:
@orangemocha the conclusion seems to be "add mkstemp refactoring the existing mkdtemp"? This is what this PR is doing.

ralt commented:
@saghul I've added the `O_CLOEXEC` usage when possible, and also added a win version. Can you run it on your CI and tell me what I messed up?

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,21 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  no_cloexec_support = 1;
+#endif
+
+  if (no_cloexec_support == 1) {
```
This detection code is not going to fly, I'm afraid. glibc added mkostemp in version 2.7, but we support 2.5 (can you configm @bnoordhuis?).

I checked a bit and turns out that some BSDs got mkostemp recently (FreeBSd 10, OpenBSD 5.7) so it's probably a safer bet to use `mkstemp` and then `fcntl` to set `O_CLOEXEC` on the fd.

bnoordhuis replied:
> glibc added mkostemp in version 2.7, but we support 2.5

Correct on both accounts.  I didn't look at the pull request closely, but one solution is to look up mkostemp() or mkostemps() through dlsym(), with fallback code for when they're not available.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,21 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
 }
 
 
+static ssize_t uv__fs_mkstemp(uv_fs_t* req) {
```
this should return int, it's the fd after all.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,21 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  if (no_cloexec_support == 1) {
+    return mkostemp((char*) req->path, O_CLOEXEC) ? 0 : -1;
+  } else {
+    return mkstemp((char*) req->path) ? 0 : -1;
```
Don't return 0 or -1, the return value is the fd we are interested in.

saghul commented:
@ralt made a more thorough pass over the Unix part, take a look at the comments.

ralt commented:
@saghul thanks for the nice feedback! I've fixed the code.

I also had to fix the tests, and I couldn't use a better way to decide if an `fd` was valid or not. On linux only, I could use something like `ASSERT(fcntl(fd, F_GETFD) != -1)`, but that wouldn't work on windows, so I'm just going with `ASSERT(fd >= 0)`.

ralt commented:
@saghul I notice that in win code, there's no `uv_fs_mkdtemp`, and we're not returning anything through this, since it returns `void`.

I'll try to see how this can be fixed for `mkstemp`.

ralt commented:
Duh, I'm stupid, I just have to find a way to get `fd` here: https://github.com/libuv/libuv/pull/729/files#diff-aa0127b468d5b4716e070fce75ecdc54R2120

ralt commented:
Fixed it, you can run CI on the win part now.

ralt commented:
@bnoordhuis as far as I understand, the issue that `mkostemp` solves over `mkstemp`/`fcntl` isn't relevant in libuv since this is run in a single-threaded environment, so the "old" way is good enough. Maybe I'm missing something though.

bnoordhuis commented:
> isn't relevant in libuv since this is run in a single-threaded environment

You mean node.js?  Because libuv most certainly supports running in a multi-threaded environment and tries hard to use atomic system calls where possible.

ralt commented:
My bad, I'll see how to use your suggestion then.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,30 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  int fd;
+
+#ifdef O_CLOEXEC
+  no_cloexec_support = 1;
```
In this case we're using FD_CLOEXEC not O_CLOEXEC (which is passed to open(2)). AFAIK that is supported on all Unix platforms.

saghul replied:
See what we do in the fallback path in `uv__fs_open`. You need to take the cloexec lock and use `uv__cloexec`. Also close the fd in case of error.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,30 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+
+  if (no_cloexec_support == 1) {
+    if (fcntl(fd, F_SETFD, FD_CLOEXEC) == -1) {
+      return -errno;
```
file descriptor leak.

ralt replied:
Thanks for bearing with me, I'll fix these tonight if I can. Alon with using Ben's suggestion.

saghul replied:
No problem :-)

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,30 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  fd = mkstemp((char*) req->path);
+
+  if (fd == -1) {
+    return -errno;
```
Return -1, the erorr is taken from errno. These functions must return the value which is expected to be put in req->result or -1, with errno set.

saghul commented on test/test-fs.c:
```diff
@@ -1069,6 +1099,39 @@ TEST_IMPL(fs_mkdtemp) {
+  loop = uv_default_loop();
+
+  r = uv_fs_mkstemp(loop, &mkstemp_req1, path_template, mkstemp_cb);
+  /* There's no cross-OS way to check an fd is valid */
```
this function is called the async way, so it doesn't return the fd, thus this comment is misleading.

saghul replied:
r should be zero if no error happened.

saghul commented on test/test-fs.c:
```diff
@@ -1069,6 +1099,39 @@ TEST_IMPL(fs_mkdtemp) {
+  /* mkstemp return different values on subsequent calls */
+  ASSERT(strcmp(mkstemp_req1.path, mkstemp_req2.path) != 0);
+
+  /* Cleanup */
```
You need to close the fds you opened with `uv_fs_close`.

saghul commented:
Left another round of comments.

ralt commented:
@saghul @bnoordhuis Please see the updated code. I *should* have taken care of all the comments.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,54 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+     of 1., so it's manually checked before. */
+  if (path_length < 6 ||
+      strcmp(path + (path_length - 6), "XXXXXX"))
+    return -EINVAL;
```
This function must return the fd or -1 with errno set.

ralt replied:
Duh, sorry for this one...

saghul commented on src/unix/core.c:
```diff
@@ -55,22 +55,6 @@
 # include <sys/ioctl.h>
 #endif
 
-#if defined(__FreeBSD__) || defined(__DragonFly__)
```
Why are you moving this?

ralt replied:
To have it available in fs.C on freebsd

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,54 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+static int uv__fs_mkstemp(uv_fs_t* req) {
+  int r;
+  int (*mkostemp_function)(char*, int);
+  char *path = (char*) req->path;
```
style: declarations and assignments in different lines

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,54 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+     NULL. */
+  dlerror();
+
+  if (mkostemp_function != NULL) {
```
style: drop braces for single statement ifs

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,54 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  dlerror();
+
+  if (mkostemp_function != NULL) {
+    return mkostemp_function(path, UV__O_CLOEXEC);
```
this could still fail with EINVAL if `open` doesn't support `O_CLOEXEC`, so you need to fallback to `mkstemp`

ralt replied:
Oops, definitely meant to do that. Sorry!

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,54 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+    r = uv__close(r);
+    if (r != 0 && r != -EINPROGRESS)
+      abort();
+    r = -1;
```
maybe copy the comment from open to clarify why we only need to return -1 in this case

bnoordhuis replied:
Actually, I think there is a bug here, the same one as in `uv__fs_open()`:  this function should return 0 when `uv__close()` returns `-EINPROGRESS` because `-EINPROGRESS` is not an error, it's informational.

Perhaps we should simply change `uv__close()` to return 0 on EINTR or EINPROGRESS; there isn't really anything you need to do, or even can do, when you get that error.

saghul replied:
let's address this in a follow-up PR?

saghul commented on src/win/fs.c:
```diff
@@ -764,8 +764,64 @@ void fs__mkdir(uv_fs_t* req) {
+  DWORD error;
+  int fd;
+
+  fd = CreateFileW(req->file.pathw,
```
after getting the file handle we need to extract its "unix style" fd suing `_open_osfhandle`, have a look at how open does it on Windows.

saghul commented on src/win/fs.c:
```diff
@@ -2053,6 +2105,26 @@ int uv_fs_mkdtemp(uv_loop_t* loop, uv_fs_t* req, const char* tpl,
 }
 
 
+int uv_fs_mkstemp(uv_loop_t* loop, uv_fs_t* req, const char* tpl,
```
style: align parameters (I know the windows side is a bit inconsistent, but new code should match the style we want not the one we have :-))

saghul commented on src/win/internal.h:
```diff
@@ -302,6 +302,7 @@ int uv_translate_sys_error(int sys_errno);
  * FS
  */
 void uv_fs_init();
+typedef int (*uv__fs_mktemp_cb)(uv_fs_t* req, static const size_t num_x, WCHAR *ep);
```
can we call this `uv__fs_mktemp_func`? I don't like using "callback" here, since we already have a callback for fs requests which is something else.

saghul commented on test/test-fs.c:
```diff
@@ -1069,6 +1100,43 @@ TEST_IMPL(fs_mkdtemp) {
+  ASSERT(strcmp(mkstemp_req1.path, mkstemp_req2.path) != 0);
+
+  /* invalid template returns EINVAL */
+  ASSERT(uv_fs_mkstemp(NULL, &mkstemp_req3, "test_file", NULL) == -EINVAL);
```
check for UV_EINVAL, -EINVAL only works on unices.

saghul commented on test/test-fs.c:
```diff
@@ -1069,6 +1100,43 @@ TEST_IMPL(fs_mkdtemp) {
+  uv_fs_close(loop, &mkstemp_req1, mkstemp_req1.file, NULL);
+  uv_fs_close(NULL, &mkstemp_req2, mkstemp_req2.file, NULL);
+  uv_fs_req_cleanup(&mkstemp_req1);
+  uv_fs_req_cleanup(&mkstemp_req2);
```
can you reverser the order here? first close the fds then unlink them. Also, pass NULL as the loop on both `uv_fs_close` calls.

saghul commented:
Another round of review, thanks for your work on this!

ralt commented:
Thanks for the reviews, I appreciate them!

ralt commented:
@saghul fixed the last round of comments, hopefully I didn't forget anything.

saghul commented:
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/261/

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,70 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  dlerror();
+
+  if (mkostemp_function != NULL) {
+    r = mkostemp_function(path, UV__O_CLOEXEC);
```
`UV__O_COELEX` is not defined for all platforms, you should ifdef this block out.

ralt replied:
Shouldn't I make `UV__O_CLOEXEC` available everywhere instead?

saghul commented:
@ralt we will also need some docs.

ralt commented:
For the windows builds, I have no idea why my branch gives these errors: https://ci.nodejs.org/job/libuv+any-pr+multi/nodes=win2008r2/261/console

ralt commented:
re: docs, they seem to be fairly recent? The existing docs seem pretty bare... unless I'm missing something.

ralt commented:
https://ci.nodejs.org/job/libuv+any-pr+multi/nodes=centos5-64-gcc41/261/console seems completely unrelated to this PR

ralt commented:
And this one seems to be a temporary network error: https://ci.nodejs.org/job/libuv+any-pr+multi/nodes=ubuntu1404-32/261/console

saghul commented:
> re: docs, they seem to be fairly recent? The existing docs seem pretty bare... unless I'm missing something.

We've had them since 1.0 and we don't add any new APIs which don't have corresponding docs. Look at how `mkdtemp` is documented and follow suit. They are bare, but something is always better than nothing :-) Also mention that this was added in 1.9.0 using ".. versionchanged:".

saghul commented:
> For the windows builds, I have no idea why my branch gives these errors: https://ci.nodejs.org/job/libuv+any-pr+multi/nodes=win2008r2/261/console

That's Jenkins having a brainfart...

saghul commented:
In general, we do have some failing tests in different platforms, just look at the ones you added / see if the build is not broken (this broke OSX, for instance).

ralt commented:
re docs: look at this on github to see my reaction https://github.com/libuv/libuv/blob/v1.x/docs/src/fs.rst

I looked at the raw file and I'd like to take back what I said. :-)

saghul commented:
> re docs: look at this on github to see my reaction https://github.com/libuv/libuv/blob/v1.x/docs/src/fs.rst
> 
> I looked at the raw file and I'd like to take back what I said. :-)

Ah, that. Yeah, GitHub doesn't render rst that good... No worries!

ralt commented:
I added an `#ifdef UV__O_CLOEXEC`, this should fix most builds (including OSX). For the windows builds, we'll have to wait for jenkins to be fixed, I guess? Both versions' builds look pretty broken.

ralt commented:
I have also added the docs. (Not much was needed, really.)

saghul commented on docs/src/fs.rst:
```diff
@@ -199,6 +200,13 @@ API
+    Equivalent to :man:`mkstemp(3)`.
+
+    .. note::
+        The result can be found as a null terminated string at `req->path`.
```
Indicate that the fd is available in req->result.

saghul replied:
Also please add .. versionadded:: 1.9.0

saghul commented:
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/263/

ralt commented:
Fixed the docs.

For CI:

- [aix61-ppc64](https://ci.nodejs.org/job/libuv+any-pr+multi/263/nodes=aix61-ppc64/console) reports way too many errors to have ever worked, such as:

```
src/unix/fs.c: In function 'uv__fs_mkdtemp':
src/unix/fs.c:216:3: warning: implicit declaration of function 'mkdtemp' [-Wimplicit-function-declaration]
   return mkdtemp((char*) req->path) ? 0 : -1;
```

Is it really my PR which is breaking this?

- centos and ubuntu are breaking on `tcp-close` tests (different ones, though). Again, I'm wondering why my PR breaks this.

- smartos is breaking on this:

```
`spawn_closed_process_io` failed: exit code 134
Output from process `spawn_closed_process_io`:
Assertion failed: fd > STDERR_FILENO, file src/unix/core.c, line 471
```

Which looks... weird.

- win builds are still borked.

-----

That said, the builds reported a legitimate issue: the `mkostemp_function` variable was unused after my `#ifdef`, I've fixed that.

Maybe we can discuss on irc about CI?

saghul commented:
> Is it really my PR which is breaking this?

I don't think so. Maybe @iWuzHere or @mhdawson can help us here. Is there any additional incldue needed on AIX for `mkdtemp` and `mkstemp`?



> centos and ubuntu are breaking on tcp-close tests (different ones, though). Again, I'm wondering why my PR breaks this.
> 
> smartos is breaking on this:

These are known failures, your PR has nothing to do with it.



> win builds are still borked.

Sigh. I'll give it a shot on my Windows machine tomorrow.

> Maybe we can discuss on irc about CI?

I'm there right now.

ralt commented:
> These are known failures, your PR has nothing to do with it.

Phew.

> Sigh. I'll give it a shot on my Windows machine tomorrow.

Good luck :-)

> I'm there right now.

Sure, irc is just more convenient for longer discussions, but I'm fine with github issues.

saghul commented:
> Sure, irc is just more convenient for longer discussions, but I'm fine with github issues.

I meant that I'm in the channel at the moment, if you want to discuss anything.

ralt commented:
I've opened a separate PR (which is not based on things I've been able to test) for the AIX issue: https://github.com/libuv/libuv/pull/740

fdopen commented on src/win/fs.c:
```diff
@@ -764,8 +764,87 @@ void fs__mkdir(uv_fs_t* req) {
+
+  file = CreateFileW(req->file.pathw,
+                     NULL,
+                     NULL,
```
The second and third parameter of CreateFile are DWORDs not pointers (NULL).
And you should pass meaningful values for dwShareMode and dwDesiredAccess. `dwShareMode` should probably be `FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE` in order to be consistent with the other functions. dwDesiredAccess `GENERIC_READ | GENERIC_WRITE` or something similar.

ralt commented:
@saghul: big thanks to who @fdopen fixed the Windows side of this PR! Your testing should be much smoother now.

saghul commented:
Awesone! Lets try the CI once more, since it behaved correctly in other builds recently: https://ci.nodejs.org/job/libuv+any-pr+multi/277/

We will need a rebase later on, yes.

ralt commented:
Woo! Working on win2012 https://ci.nodejs.org/job/libuv+any-pr+multi/277/nodes=win2012r2/console

Win2008 build is still borked though.

ralt commented:
And rebased. I kept his commit as is to keep the attribution.

saghul commented on docs/src/fs.rst:
```diff
@@ -1,4 +1,3 @@
-
```
please leave this one alone

ralt replied:
Oops, damn editor...

saghul commented on docs/src/fs.rst:
```diff
@@ -199,6 +199,18 @@ API
+    .. note::
+        The created file path can be found as a null terminated string at `req->path`.
+
+    .. note::
```
Combine these into a single  note.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,74 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+
+    /* If mkostemp() returns EINVAL, it means the kernel doesn't
+       support O_CLOEXEC, so we just fallback to mkstemp() below. */
+    if (r == -1 && errno != -EINVAL)
```
here, and above, remember if mkostemp was available and worked using s static variable, to avoid doing the dance every time

saghul commented on src/win/fs.c:
```diff
@@ -803,15 +802,10 @@ void fs__mkdtemp(uv_fs_t* req) {
-      break;
-    } else if (errno != EEXIST) {
-      SET_REQ_RESULT(req, -1);
+    if (callback(req, num_x, ep)) {
```
please call it func, not callback

saghul replied:
style: drop braces for single statement ifs

saghul commented on src/win/fs.c:
```diff
@@ -821,6 +815,84 @@ void fs__mkdtemp(uv_fs_t* req) {
   }
 }
 
+static int fs__mkdtemp_callback(uv_fs_t* req, size_t num_x, WCHAR* ep) {
```
don't use callback as a name

saghul commented:
Left some style related comments. Other than that, tests are :green_heart: Can you also please rebase? FYI, you might run into a conflict near the definition of UV__O_CLOEXEC.

ralt commented:
For the win changes, I have 2 choices:

- Add a new commit to fix your notes
- Rebase the changes into existing commits

2. sounds like the best option, but I'm afraid I'm going to lose the attribution of @fdopen's commit. If I can't get to it, I prefer going with the first option.

I'll give it a try and rebase against existing 1.x, thanks for the comments!

saghul commented:
> And rebased. I kept his commit as is to keep the attribution.

The commits need to be squeashed into one, as it will land as a single commit, so feel free to credit him in the commit message.

ralt commented:
Oh, ok. Will do then.

cjihrig commented on src/unix/fs.c:
```diff
@@ -217,6 +217,74 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+#ifdef UV__O_CLOEXEC
+  int (*mkostemp_function)(char*, int);
+#endif
+  char *path;
```
Can you change the star alignment. `char* path`

cjihrig commented on src/unix/fs.c:
```diff
@@ -217,6 +217,74 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+       2. open() didn't support O_CLOEXEC
+     We want to avoid going to the fallback path in case
+     of 1., so it's manually checked before. */
+  if (path_length < 6 ||
```
Maybe the 6 should be made into a constant.

bnoordhuis replied:
Could be done like this:
```c
static const char pattern[] = "XXXXXX";
// ...
if (path_length < sizeof(pattern) - 1 || /* ... */) {
```

cjihrig commented on src/unix/fs.c:
```diff
@@ -217,6 +217,74 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  if (path_length < 6 ||
+      strcmp(path + (path_length - 6), "XXXXXX")) {
+    errno = EINVAL;
+    return -1;
```
Can you document why -1 is returned instead of the error.

ralt commented:
@saghul I've amended with your suggestions, squashed everything into a single commit, and rebased against upstream/v1.x. I suggest running CI at least one last time if this is good enough to merge.

fdopen commented on src/unix/fs.c:
```diff
@@ -217,6 +217,82 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+
+#ifdef UV__O_CLOEXEC
+  if (no_cloexec_support == 0) {
+    *(int**)(&mkostemp_function) = dlsym(RTLD_DEFAULT, "mkostemp");
```
mkostemp can be a macro (or hidden by a macro under certain conditions). It's eg. sometimes the case on 32-bit systems, where mkostemp is redirected to mkostemp64, in order to pass O_LARGEFILE to open.

The presence of mkostemp should be  checked at compile time, not at runtime. Otherwise you risk to call the wrong function.

saghul replied:
Ouch. We do runtime checks because we allow for running libuv on systems other than the one where it was compiled.

saghul replied:
I guess we can just ignore `mkostemp` if we can't reliably load it at runtime. @bnoordhuis WDYT?

bnoordhuis replied:
I doubt you'll find a libc that has a mkostemp64 symbol but not a mkostemp one so I don't think it's going to be an issue in practice.  If you want to play it safe, try looking up mkostemp64 before mkostemp, although I'd probably just pass O_LARGEFILE explicitly (good point, that.)

ralt replied:
So, just look for `mkostemp64` and pass `O_LARGEFILE` to it, completely ignoring `mkostemp`?

saghul replied:
I think so. @bnoordhuis can you confirm?

bnoordhuis replied:
Ah, not quite.  If you find a `mkostemp64` symbol, use that, else look for `mkostemp`, and always pass `O_LARGEFILE`.

ralt replied:
@bnoordhuis so it's possible to `mkostemp64` to not exist while `mkostemp` does? Looking at the source:

- `mkostemp` defines itself as an alias of `mkostemp64` is `O_LARGEFILE == 0` or `O_LARGEFILE` doesn't exist
- `mkostemp64` is only defined when `O_LARGEFILE` exists _and_ `O_LARGEFILE != 0`

So passing `O_LARGEFILE` to `mkostemp` sounds useless since it won't exist, otherwise we're using `mkostemp64`.

Or am I missing something?

bnoordhuis replied:
Different libcs implement `mkostemp` differently.  So yes, `mkostemp64` need not exist.

ralt replied:
Ah, didn't think about other libc. Thanks! I'll do that when I get some time.

saghul commented:
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/284/

Before we land this (is tests are green) we need 2 more things:

* address this: https://github.com/libuv/libuv/pull/729#discussion_r56497617
* please don't call the processor "callback", that code already has a callback, and it's a completely different thing

saghul commented:
@ralt Any chance you can finish this soon? (no pressure, just trying to see if we should wait for the next release to land it instead of 1.9.0)

ralt commented:
@saghul life is busy these days and I don't know when I'll have time so don't wait for me.

saghul commented:
No worries Florian, and thanks for your work! It will make it to the next
one.
On Apr 3, 2016 16:35, "Florian Margaine" <notifications@github.com> wrote:

> @saghul <https://github.com/saghul> life is busy these days and I don't
> know when I'll have time so don't wait for me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/729#issuecomment-205040592>
>

ralt commented:
@saghul @bnoordhuis I've pushed the new code handling `mkostemp64` and `O_LARGEFILE`.

bnoordhuis commented on docs/src/fs.rst:
```diff
@@ -199,6 +200,16 @@ API
+        The created file path can be found as a null terminated string at `req->path`.
+        The file descriptor can be found as an integer at `req->result`.
+
+    .. versionadded:: 1.9.0
```
To whoever lands this: should be updated to the current version.

ralt replied:
Replaced with 1.10.0

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -217,6 +217,88 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  int (*mkostemp_function)(char*, int);
+  static int no_cloexec_support;
+#endif
+  char *path;
```
Style nit: `char* path`

ralt replied:
Fixed

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -217,6 +217,88 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+
+      /* If mkostemp() returns EINVAL, it means the kernel doesn't
+         support O_CLOEXEC, so we just fallback to mkstemp() below. */
+      if (r == -1 && errno != -EINVAL)
```
Check `errno == EINVAL`, not the negated version of EINVAL, here and below.

ralt replied:
Err, sorry, I don't get it. If I check `errno == EINVAL`, then it must not return at this point, no?

ralt replied:
Or did you mean `if (r == -1 && errno != EINVAL)`?

bnoordhuis replied:
Sorry, that's what I mean; the important part was not checking the negated version. :-)

ralt replied:
Phew, I thought my sanity was gone. Pushed.

bnoordhuis commented on src/unix/fs.c:
```diff
@@ -217,6 +217,88 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+  char *path;
+  int path_length;
+
+  path = (char*) req->path;
```
I'm thinking that maybe we should change back `uv_fs_t.path` to a `char*`.  We can do that without API or ABI breakage.  In-place modification of a `const char*` is surprising, to say the least.

Also, I expect you need to update `uv_fs_req_cleanup()` to special-case `UV_FS_MKSTEMP` requests.

ralt replied:
This is a problem already existing with `uv_fs_mkdtemp`, no? Anyway, I made it a `char*` and the tests pass. If you're saying the ABI/API is fine, then we're good to go?

I've also fixed `uv_fs_req_cleanup`.

ralt commented:
@bnoordhuis all your points should have been fixed.

saghul commented on src/unix/fs.c:
```diff
@@ -217,6 +217,89 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+
+      /* We set the static variable so that next calls don't even
+         try to use mkostemp. */
+      if (errno == EINVAL)
```
isn't this always true?

ralt replied:
Indeed, it is!

saghul commented on src/win/fs.c:
```diff
@@ -763,9 +763,8 @@ void fs__mkdir(uv_fs_t* req) {
-
 /* OpenBSD original: lib/libc/stdio/mktemp.c */
-void fs__mkdtemp(uv_fs_t* req) {
+static void fs__mktemp(uv_fs_t* req, uv__fs_mktemp_func callback) {
```
Please, **please** do not name that *callback*.

saghul commented:
Some minor comments otherwise LGTM if the CI is :green_heart: 

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/37/

ralt commented:
@saghul fixed your 2 comments, and CI looks as green as it can be.

saghul commented:
Rubber-Stamp LGTM, great work! @bnoordhuis, one last look?

ralt commented:
Any news? I see that the branch now has conflicts, is that something I should be worried about?

saghul commented:
Oops, this fell through the cracks :-( Can you please rebase it?

ralt commented:
No problem, I've rebased on top of v1.x. The conflict was in the include, easily fixable.

saghul commented on src/win/internal.h:
```diff
@@ -302,6 +302,7 @@ int uv_translate_sys_error(int sys_errno);
  * FS
  */
 void uv_fs_init();
+typedef int (*uv__fs_mktemp_func)(uv_fs_t* req, size_t num_x, WCHAR *ep);
```
just noticed that we don't use this outside fs.c, so can you mode it there?

ralt replied:
Just at the top of the file? There's no other typedef in there, so I'm not sure where this should go.

saghul replied:
move this to fs.c, just before `fs__mktemp`.

saghul commented:
I ran another CI build:

and now FreeBSD fails (due to the lack of O_LARGEFILE):

~~~~
src/unix/fs.c:268:51: error: use of undeclared identifier 'O_LARGEFILE'
      r = mkostemp_function(path, UV__O_CLOEXEC | O_LARGEFILE);
                                                  ^
~~~~

Also some warnings:

~~~~
src/unix/fs.c:246:19: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
  if (path_length < sizeof(pattern) - 1 ||
      ~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~
src/unix/fs.c:1049:3: warning: assigning to 'char *' from 'const char *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
  PATH;
  ^~~~
src/unix/fs.c:81:17: note: expanded from macro 'PATH'
      req->path = path;         
~~~~

ralt commented:
So, I've converted the path to `char*` instead of `const char*`, and it's a pretty big change:

- Many function signatures in `include/uv.h` change
- Many function signatures in `src/win/fs.c` change
- Many function signatures in `src/unix/fs.c` change
- Many tests have to be fixed to pass `char*` instead of `const char*`

I'm fine with doing the change, the compiler is mostly driving me. Just wanted to make sure it's ok.

saghul commented:
@bnoordhuis you suggested doing away with the `const`. Maybe it's better to keep to avoid the churn?

ralt commented:
it's also possible that I misunderstood what's wanted, since @bnoordhuis said there would be no API break.

saghul commented:
@bnoordhuis ping?

saghul commented:
@ralt If you still have the time, let's try to bring this home. Ben is quite busy these days, so let's leave the const alone and cast it away as necessary. Removing it can always be done as a follow-up patch later on.

You probably need to rebase, sorry about that!

saghul commented:
Ping? @ralt If you no longer have the time / inclination to move this forward, would you mind if somebody else does?

(no real preassure, just trying to schedule a couple of things here)

ralt commented:
Oops, sorry, I planned on fixing the conflicts... I'll try to do it some time this week or the next one.

saghul commented:
Great, no problemo then!

ralt commented:
@saghul rebased on top of 1.x.

saghul commented on include/uv.h:
```diff
@@ -1113,7 +1114,7 @@ struct uv_fs_s {
   uv_fs_cb cb;
   ssize_t result;
   void* ptr;
-  const char* path;
```
undo this change, it creates too much churn. cast away the const where necessary (which should be a single place IIRC)

ralt replied:
Ok, it wasn't necessary to cast it.

saghul commented on src/unix/fs.c:
```diff
@@ -235,6 +235,93 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
+   */
+  if (r >= 0 && uv__cloexec(r, 1) != 0) {
+    r = uv__close(r);
+    if (r != 0 && r != -EINPROGRESS)
```
remove the EINPROGRESS check, we fixed that by adding the check inside uv__close.

saghul commented on src/unix/fs.c:
```diff
@@ -1152,6 +1240,21 @@ int uv_fs_mkdtemp(uv_loop_t* loop,
+                  const char* tpl,
+                  uv_fs_cb cb) {
+  INIT(MKSTEMP);
+  req->path = uv__strdup(tpl);
```
you'll want to cast away the const here

saghul commented on test/test-fs.c:
```diff
@@ -1069,6 +1100,69 @@ TEST_IMPL(fs_mkdtemp) {
+  ASSERT(uv_fs_mkstemp(NULL, &mkstemp_req3, "test_file", NULL) == UV_EINVAL);
+
+  do {
+    wr = write(mkstemp_req1.result, "X", 1);
```
what is this supposed to be testing? if you can actually write to the file? if so, use `uv_fs_write`, since this is not portable.

ralt replied:
Indeed, and fixed. Thanks.

saghul commented:
@ralt thanks! left another few comments. We're close!

ralt commented:
Hey @saghul I finally got back to this, thanks for your patience!

I've fixed your comments and updated the PR.

sam-github commented:
@ralt Is this still moving forward? do you want it adopted?

ralt commented:
@sam-github as far as I know, the PR is ready. I'm not sure what else needs to be done.

I haven't particularly paid attention to getting it merged though, but in and of itself, there's not much to do.

cjihrig commented:
It looks like this would at least need a rebase to get things started again.

bshm commented:
@sam-github @cjihrig I'd like to adopt this PR. Is there a way to do it on this PR or should a create a new one and cross link it?

ralt commented:
If needed: anyone has my full permission to take over this PR.

cjihrig commented:
@bshm I think opening a new PR would be best. I'll close this one out.
